contracts/Governance/Timelock.sol

```diff
@@ -1,123 +1,177 @@
+        emit NewMinimumExecutionDelaySet(minimumExecutionDelay);
     }
 
-    function acceptAdmin() external {
-        require(msg.sender == pendingAdmin, "Timelock::acceptAdmin: Call must come from pendingAdmin.");
-        admin = msg.sender;
-        pendingAdmin = address(0);
+    function setMaximumExecutionDelay(uint256 _maximumExecutionDelay) public onlyOwner {
+        //todo ag tests
+        require(_maximumExecutionDelay >= minimumExecutionDelay, "Timelock: Maximum execution delay cannot be lesser than minimum execution delay.");
+
+        maximumExecutionDelay = _maximumExecutionDelay;
 
-        emit NewAdmin(admin);
+        emit NewMaximumExecutionDelaySet(_maximumExecutionDelay);
     }
 
-    function setPendingAdmin(address pendingAdmin_) external {
-        require(msg.sender == address(this), "Timelock::setPendingAdmin: Call must come from Timelock.");
-        pendingAdmin = pendingAdmin_;
+    function setExecutionGracePeriod(uint256 _executionGracePeriod) public onlyOwner {
+        //todo ag tests
+        executionGracePeriod = _executionGracePeriod;
 
-        emit NewPendingAdmin(pendingAdmin);
+        emit NewExecutionGracePeriodSet(_executionGracePeriod);
     }
 
-    function queueTransaction(
-        address target,
-        uint256 value,
-        string memory signature,
-        bytes memory data,
-        uint256 eta
-    ) external returns (bytes32) {
-        require(msg.sender == admin, "Timelock::queueTransaction: Call must come from admin.");
-        require(eta >= getBlockTimestamp() + delay, "Timelock::queueTransaction: Estimated execution block must satisfy delay.");
+    function setExecutionDelay(uint256 _executionDelay) public onlyOwner {
+        require(_executionDelay >= minimumExecutionDelay, "Timelock: Execution delay must exceed minimum execution delay.");
+        require(_executionDelay <= maximumExecutionDelay, "Timelock: Execution delay must not exceed maximum execution delay.");
 
-        bytes32 txHash = keccak256(abi.encode(target, value, signature, data, eta));
-        queuedTransactions[txHash] = true;
+        executionDelay = _executionDelay;
 
-        emit QueueTransaction(txHash, target, value, signature, data, eta);
-        return txHash;
+        emit NewExecutionDelaySet(executionDelay);
     }
 
-    function cancelTransaction(
-        address target,
-        uint256 value,
-        string memory signature,
-        bytes memory data,
-        uint256 eta
-    ) external {
-        require(msg.sender == admin, "Timelock::cancelTransaction: Call must come from admin.");
+    function queueTransactionsBatch(Transaction[] memory transactions, uint256 executionStartTimestamp) external onlyProposer returns (bytes32) {
+        //todo ag hmm... do we really want it this way? we definitely want 0 delay on local deployment
+        require(executionStartTimestamp >= block.timestamp + executionDelay, "Timelock: Estimated execution time must satisfy delay.");
```

Comment:
let's add another requirement for the `transactions` array to have at least one value